### DISCUSSION TITLE
U8G2_ST7567_ENH_DG128064I_1_HW_I2C

### DISCUSSION THREAD

**Discussion by mvromancini** [olikraus/u8g2#2134] [Q&A]
Hello, I've been using the U8G2 for some time and I haven't had any problems with the constructions of the SSD 1106 and even 1309 128x64 I2C display line. but I'm having difficulties with the display U8G2_ST7567_ENH_DG128064I_1_HW_I2C u8g2(U8G2_R2,U8X8_PIN_NONE); it doesn't work in 1_WH_I2C mode only in SW_I2C could you help me. OBS I use the entire SSD I2C display line without problems with the U8G2.H in HW mode. but for reasons of size and cost I am having to use another display in the case of the GM12864-59N VER 2.0 IIC and I am not getting success in working it in my project in HW_i2C mode.

**Comment by olikraus** [comment]
If there is no wiring issue, you could lower the bus frequency: https://github.com/olikraus/u8g2/wiki/u8g2reference#setbusclock

**Reply by mvromancini** [reply]
Hi Olikraus. I defined that this display does not work for my use nor for those who use other i2c devices in the project, because even the code that the Chinese offers only works on the SCl SDA pins, which is HW ..(5.4 or 21.20 ) in this case of the mega 2560 and NANO 328 boards.... I really appreciate your attention for having answered me promptly. Much obliged!!

Chinese page link where there is the code that shows that it only works in SW.

http://www.yizu.org/archives/814/

**Reply by olikraus** [reply]
You need to connect your display according to the following pins:

Board | I2C/TWI pins
-- | --
UNO, Ethernet | A4 (SDA), A5 (SCL)
Mega2560 | 20 (SDA), 21 (SCL)
Leonardo | 20 (SDA), 21 (SCL), SDA1, SCL1

Add 10K pullups. Once done the HW_I2C constructor should work.

**Reply by mvromancini** [reply]
Hello. Oliver . It took me a while but I did all the tests possible and with several displays and different manufacturers. with pull up resistors, changing frequency, changing hardware platform etc. unfortunately this constructor "U8G2_ST7567_ENH_DG128064I_1_HW_I2C u8g2(U8G2_R2,U8X8_PIN_NONE);" it doesn't work in HW_I2C mode, but going through all the constructors I found this "U8G2_ST7567_HEM6432_1_HW_I2C" that worked but in the proper size proportions. I'm still testing all the ST7567 constructors to find one that works!

**Reply by olikraus** [reply]
Note: The display controller ST7567 might be the same, but still there are a lot of other differences between displays with the same display controller. As a consequence a constructor may not work at all. We need the display datasheet to crosscheck whether it fits to a given constructor.

**Reply by mvromancini** [reply]
worked just fine.!!!   I found ! by comparison with the .C files of the constructors that worked and I put the line that was missing in relation to the constructor in question.. and it worked just fine.

this one, as I explained works but with the screen in the proper proportions "64x32"

uint8_t u8x8_d_st7567_hem6432(u8x8_t *u8x8, uint8_t msg, uint8_t arg_int, void *arg_ptr)
{
  uint8_t x, c;
  uint8_t *ptr;
  switch(msg)
  {
    case U8X8_MSG_DISPLAY_SETUP_MEMORY:
      u8x8_d_helper_display_setup_memory(u8x8, &u8x8_st7567_hem6432_display_info);
      **u8x8->i2c_address = 0x07e;**  /* issue 1159, use different i2c address */
      break;
include this line in the constructor and then it worked
      **u8x8->i2c_address = 0x07e;**  /* issue 1159, use different i2c address */

uint8_t u8x8_d_st7567_enh_dg128064i(u8x8_t *u8x8, uint8_t msg, uint8_t arg_int, void *arg_ptr)
{
  switch(msg)
  {
    case U8X8_MSG_DISPLAY_SETUP_MEMORY:
      u8x8_d_helper_display_setup_memory(u8x8, &u8x8_st7567_enh_dg128064i_display_info);
      **u8x8->i2c_address = 0x07e;**  /* issue 1159, use different i2c address */
      break;

**Reply by olikraus** [reply]
What exactly was missing?

**Reply by mvromancini** [reply]
uint8_t u8x8_d_st7567_enh_dg128064i(u8x8_t *u8x8, uint8_t msg, uint8_t arg_int, void arg_ptr)
{
switch(msg)
{
case U8X8_MSG_DISPLAY_SETUP_MEMORY:
u8x8_d_helper_display_setup_memory(u8x8, &u8x8_st7567_enh_dg128064i_display_info);
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
break;

u8x8_d_helper_display_setup_memory(u8x8, &u8x8_st7567_hem6432_display_info);
**u8x8->i2c_address = 0x07e; / issue 1159, use different i2c address /**
break;

The line with the I2C address

**Reply by mvromancini** [reply]
Thank you, friend, for giving me the attention to answer all my questions. and I consider the problem solved and here is the correction tip for the constructors of the St7567 line in I2C that do not work in HW_i2C!

**Reply by olikraus** [reply]
No problem, note that there 8s also a high level function to change the i2c address.

**Reply by mvromancini** [reply]
Yes, after picking it up and analyzing it, it was clear that you have it and it was worth getting to know your project in more depth, which is very good,,,keep it up always!!!!

**Reply by thinhdinh25** [reply]
I do not really understand what changes you have made. Can u explain
